### ISSUE TITLE
Unable to acquire new refresh token

### ISSUE THREAD

**Issue by travisnielsen** [Azure/go-autorest#196]
I'm unsure if this is the right place to log this issue, but I'm experimenting with the recently announced [configurable token lifetimes](https://docs.microsoft.com/en-us/azure/active-directory/active-directory-configurable-token-lifetimes) feature in AAD. The purpose is to enforce an idle timeout for Kubernetes (kubectl) users so that they must re-authenticate to Azure after a given amount of idle time. In our scenario, we have OIDC enabled in Kubernetes and we're finding developers are sharing their config files (despite being told not to). We're looking to (partially) mitigate this by enforcing more frequent authentications.

To achieve this, I'm setting the "Refresh Token Max Inactive Time" property in a custom policy and linking it to the AAD object that represents the k8s API. When this policy is active, kubectl is unable to re-authenticate the user and acquire a new refresh token after the idle timeout period is reached:

`E1116 09:51:11.431092    3771 azure.go:126] Failed to acquire a token: refreshing the expired token: refreshing token: adal: Refresh request failed. Status Code = '400'
Unable to connect to the server: acquiring a token for authorization header: refreshing the expired token: refreshing token: adal: Refresh request failed. Status Code = '400'`

I suppose this could be a issue related to AAD, the [ADAL Go library](https://github.com/cosmincojocar/adal), or even the kubectl plug-in for Azure. I'm looking for guidance as to where this issue would best be reported and what further information I can provide.

**Comment by mcardosos** [comment]
@khenidak
Do you know how kubernetes is calling AAD in this scenario?
@travisnielsen
Is kubectl using go-autorest or the ADAL Go library?

**Comment by khenidak** [comment]
Here: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/plugin/pkg/client/auth/azure/azure.go 

I think ADAL (and kubectl’s plug-in is based on it) don’t do retry quite right.

**Comment by travisnielsen** [comment]
Yes. kubectl is using go-autorest (see Kal's link above). I'll try to get more details on how the call is being executed and will move this to the ADAL team.

**Comment by vladbarosan** [comment]
Hi @travisnielsen , you have more details for us on this ? thanks

**Comment by travisnielsen** [comment]
Here are some more details. Versions: Kubectl: 1.9.0, Kubernetes: 1.8.4. RBAC functionality with AAD authentication works fine. I then created an [Azure AD Token Lifetime Policy](https://docs.microsoft.com/en-us/azure/active-directory/active-directory-configurable-token-lifetimes) with the following settings:
```
"AccessTokenLifetime":"0.00:10:00" - lifetime of the access token (10 mins)
"MaxInactiveTime":"00.00:15:00" - max time a refresh token can be idle (15 mins)
"MaxAgeMultiFactor":"until-revoked" - overall lifetime of a refresh token since originally issued
"MaxAgeSingleFactor":"until-revoked" - overall lifetime of a refresh token since originally issued
```
After linking this policy to the AAD application object that represents the API server and re-configuring the kubectl AAD authentication, I confirmed the access token lifetime was indeed 10 minutes as specified in the policy. Just after 10 minutes had passed, I ran kubectl again and I was able to get another access token (via the refresh token). This was to be expected because I was within the refresh token inactive time window (15 minutes).

I then waited until 15 minutes had passed and ran kubectl again. Because the refresh token inactivity window had elapsed, my expectation was that the device authentication flow would be re-started. However, kubectl reports the following error:
```
kubectl get nodes --user alice
E0110 21:47:59.926060    6352 azure.go:126] Failed to acquire a token: refreshing the expired token: refreshing token: adal: Refresh request failed. Status Code = '400'. Response body: {"error":"invalid_grant","error_description":"AADSTS70002: Error validating credentials. AADSTS70008: The refresh token has expired due to inactivity. The token was issued on 2018-01-11T03:06:26.9607996Z and was inactive for 00:15:00.\r\nTrace ID: 77c5eb54-d0a4-4926-a1bc-2c9f76ae0100\r\nCorrelation ID: eb3a75d5-50be-417d-ba9e-99975220e808\r\nTimestamp: 2018-01-11 03:48:00Z","error_codes":[70002,70008],"timestamp":"2018-01-11 03:48:00Z","trace_id":"77c5eb54-d0a4-4926-a1bc-2c9f76ae0100","correlation_id":"eb3a75d5-50be-417d-ba9e-99975220e808"}
```

**Comment by vladbarosan** [comment]
Synced offline with @travisnielsen , this is expected behavior. will be making the change in Kubernetes to handle expired refresh tokens
